src/net/stackerdb/mod.rs

```diff
@@ -0,0 +1,133 @@
+/// `read` on a StackerDB is the act of loading one chunk from the node's local replica.  Reading
+/// and writing a single chunk on one node is atomic.  StackerDB replication proceeds in a
+/// store-and-forward manner -- newly-discovered chunks are stored to the node's local replica and
+/// broadcast to a subset of neighbors who also replicate the given StackerDB.
+///
+/// Each chunk has an associated Lamport clock, and an associated public key hash used to
+/// authenticate writes.  The Lamport clock is used to identify the latest version of a chunk --
+/// a node will replace an existing but stale copy of a chunk with a newly-downloaded chunk if its
+/// Lamport clock has a strictly higher value.  The chunk's metadata -- its ID, Lamport clock, and
+/// data hash -- must be signed by the chunk's public key hash's associated private key in order to
+/// be stored.  The chunks themselves are ordered byte sequences with no mandatory internal
+/// structure.
+///
+/// StackerDB state is ephemeral.  Its contents are dropped at the start of every reward cycle.
+/// Endpoints must re-replicate data to the StackerDB if they wish to keep it online.  In doing so,
+/// the set of StackerDBs is self-administrating -- a node will only store state for active
+/// StackerDBs.
+///
+/// ## Control Plane
+///
+/// The smart contract to which a StackerDB is bound controls how many chunks the DB has, who can
+/// write to which chunks (identified by public key hash), how big a chunk is, and how often a
+/// chunk can be written to (in wall-clock time).  This smart contract is queried once per reward cycle
+/// in order to configure the database.  The act of configuring the re-configuring the database
+/// is also the act of dropping and reinstantiating it.
+///
+/// Applications that employ StackerDBs would deploy one or more smart contracts that list out
+/// which users can store data to the StackerDB replica, and how much space they get.
+///
+/// ## Replication Protocol
+///
+/// StackerDB replication proceeds in a three-part protocol: discovery, inventory query, and
+/// chunk exchange.  The discovery protocol leverages the Stacks node's neighbor-walk algorithm to
+/// discover which StackerDBs other nodes claim to replicate.  On receipt of a `Handshake` message,
+/// a StackerDB-aware node replies with a `StackerDBHandshakeAccept` message which encodes both the
+/// contents of a `HandshakeAccept` message as well as a list of local StackerDBs (identified by
+/// their smart contracts' addresses).  Upon receipt of a `StackerDBHandshakeAccept`, the node
+/// stores the list of smart contracts in its `PeerDB` as part of the network frontier state.  In
+/// doing so, nodes eventually learn of all of the StackerDBs replicated by all other nodes.  To
+/// bound the size of this state, the protocol mandates that a node can only replicate up to 256
+/// StackerDBs.
```

Comment:
Can you add a reference to where the handshake is implemented (i.e., `net::chat::handle_handshake`)?

Reply:
Fixed